#477 - part 1 - Refactoring structure stage date on Application, review and assignment

nmadruga opened the pull request:
Partial fix for #477 

The idea is to have the `stageDate` available for Review and Application when showing the history, but just prefer to break-down this PR which is doing a small refactor so it doesn't mix with specific changes to show the History modal next (part 2) 

### Linked PRs
[#477F-part1](https://github.com/openmsupply/application-manager-server/pull/412)

### Changes
- A lot of files with only variable changes....
- Changes on types was to unify a little how we keep stages information:
   - Removed type `StageAndStatus`
   - Replaced  `StageAndStatus` with `ApplicationStage` in FullStructure of application
   - Replace `Review.stage` from `ApplicationStage` to new type `ReviewStage` and move status to inside of that
   - Extends `ReviewStage` and `ApplicationStage` from same type `StageDetails`
- Renamed in `FullStructure` (for Application) field `current` to `currentStage` it keeps both dates `currentDate` for this stage and `statusCreatedDate` for the current `status` property.
- Fix on `ReviewSectionRowLastActionDate` to show correct date related to the status (instead of stage as it was before)

andreievg commented on src/components/Review/ReviewSectionRowLastActionDate.tsx:
```diff
@@ -16,15 +16,15 @@ const ReviewSectionRowLastActionDate: React.FC<ReviewSectionComponentProps> = ({
           <LastDate
             title={strings.ACTION_DATE_REVIEW_STARTED}
-            indicator={getSimplifiedTimeDifference(thisReview?.timeStatusCreated)}
+            indicator={getSimplifiedTimeDifference(thisReview?.stage.statusCreatedDate)}
```
i think this is different now, as previously it was date review was started, and I think stage.statusCreatedDate is when application jumped to new stage

andreievg replied:
Ahh i see you've changed to putting statusCreated together with stage, which is a bit confusing for me, I think the stage no ReviewDetails should be explicitly 'applicationStage' -> to avoid the confusion, and maybe keep review status related field within the review, rather then stage ?

nmadruga replied:
Yes... I can see it might be a bit confusing indeed. Before we had a composed property:
```

nmadruga replied:
Will be changing this to the above suggestion, using fields for times more aligned with how is stored in `review` table:
interface StageAndStatus {
  stage: StageDetails
  status: ApplicationStatus
  timeStageCreated: Date
  timeStatusCreated: Date 
}
```

andreievg commented on src/utils/graphql/queries/getReviewInfo.query.ts:
```diff
@@ -6,13 +6,19 @@ export default gql`
     reviewAssignments(condition: { applicationId: $applicationId }, orderBy: TIME_UPDATED_DESC) {
       nodes {
         id
-        levelNumber
```
I think this mike cause a problem ?

nmadruga replied:
It's a duplication : )

andreievg commented on src/utils/hooks/useLoadApplication.ts:
```diff
@@ -99,15 +107,14 @@ const useLoadApplication = ({ serialNumber, networkFetch }: UseGetApplicationPro
+        createdDate: statusHistoryTimeCreated,
         status: status as ApplicationStatus,
-        date: statusHistoryTimeCreated,
+        statusCreatedDate: stageHistoryTimeCreated,
```
is this correct ? there is no graphQL query for this and stage history time created relates to creation time of stage not latest status. similar to comment about review, maybe we should seperate currentStage state from currentStatus state

nmadruga replied:
Good catch, this is wrong. It should be the opposite way. Changing this now

andreievg replied:
What about the graphQL part of the comment? I can't see where it's queried

nmadruga replied:
Oh, I didn't see this one. Thanks! Fixing now : )

andreievg commented on src/utils/graphql/queries/getReviewInfo.query.ts:
```diff
@@ -6,13 +6,19 @@ export default gql`
+          title
+          colour
+        }
+        stageDate
```
was wondering is this the only 'data' addition for this PR ?

nmadruga replied:
Yes. Pretty much, I have had to refactor things to also avoid using different types to keep the same data:
* Removed type StageAndStatus
* Replaced StageAndStatus with ApplicationStage in FullStructure of application
* Replace Review.stage from ApplicationStage to new type ReviewStage and move status to inside of that

nmadruga commented:
@andreievg any further comments on this PR? I would like to continue dealing with problem/suggestion on #477 - part 2.

nmadruga commented:
Hi @andreievg I have changed this to be back on using the `current` field (instead of `status` inside `stage` field) and updated the fields related to timestamps on `StageAndStatus`, `AssignmentStageAndStatus` and `ReviewStageAndStatus`, hope this is clearer now. 

I'm currently working on a required change for #477-part2 to show reviewResponses in order, will commit my changes to the following PR which will have this PR already up-to-date.

If you have time today to review this one that will be great and probably practical to get more things included in either `development` or `master` branch. Depending on how we choose to proceed.
